kernel/kernel-default/src/main/java/io/delta/kernel/client/DefaultJsonHandler.java

```diff
@@ -0,0 +1,187 @@
+                // initialize the next file reader or return false if there are no more files to
+                // read.
+                try {
+                    if (currentFileReader == null ||
+                        (nextLine = currentFileReader.readLine()) == null) {
+
+                        tryOpenNextFile();
+                        if (currentFileReader != null) {
+                            nextLine = currentFileReader.readLine();
+                        }
+                    }
+                }
+                catch (IOException ex) {
+                    throw new RuntimeException(ex);
+                }
+
+                return nextLine != null;
+            }
+
+            @Override
+            public FileDataReadResult next()
+            {
+                if (nextLine == null) {
+                    throw new NoSuchElementException();
+                }
+
+                List<Row> rows = new ArrayList<>();
+                int currentBatchSize = 0;
+                do {
+                    // hasNext already reads the next one and keeps it in member variable `nextLine`
+                    rows.add(parseJson(nextLine, physicalSchema));
+                    nextLine = null;
+                    currentBatchSize++;
+                }
+                while (currentBatchSize < maxBatchSize && hasNext());
+
+                ColumnarBatch batch = new DefaultRowBasedColumnarBatch(physicalSchema, rows);
+                Row scanFileRow = currentFile.getScanFileRow();
+                return new FileDataReadResult()
+                {
+                    @Override
+                    public ColumnarBatch getData()
+                    {
+                        return batch;
+                    }
+
+                    @Override
+                    public Row getScanFileRow()
+                    {
+                        return scanFileRow;
+                    }
+                };
+            }
+
+            private void tryOpenNextFile()
+                throws IOException
+            {
+                Utils.closeCloseables(currentFileReader); // close the current opened file
+                currentFileReader = null;
+
+                if (fileIter.hasNext()) {
+                    currentFile = fileIter.next();
+                    FileStatus fileStatus =
+                        Utils.getFileStatus(currentFile.getScanFileRow());
+                    Path filePath = new Path(fileStatus.getPath());
+                    FileSystem fs = filePath.getFileSystem(hadoopConf);
+                    FSDataInputStream stream = null;
+                    try {
+                        stream = fs.open(filePath);
+                        currentFileReader = new BufferedReader(
+                            new InputStreamReader(stream, StandardCharsets.UTF_8));
+                    } catch (Exception e){
+                        Utils.closeCloseables(stream); // close it avoid leaking resources
```

Comment:
propagate the exception?

Reply:
`Utils.closeCloseables` already throws an unchecked exception. 

Related to this: The iterator interface needs to have some kind of checked exception throwing, but we can't override the interface. Added this to the list. We may need to revisit this. Unchecked exceptions are not always good.

Reply:
Isn't the original exception `e` suppressed here though?

Reply:
+1 on the iterator interface

Reply:
No its not.

```
            if (exception == null) {
                    exception = new RuntimeException(ex);
                }
                else {
                    exception.addSuppressed(ex);
                }
```

We keep the first exception (wrap it in RuntimeException because of the no-checked exceptions). Subsequent exceptions are added to suprrssed. they are still available, but won't be printed.

Reply:
I mean whatever exception caused us to reach the catch block here at all. Not sure if I'm missing something obvious

Reply:
If we fail to open the file or something and `stream = null` we just continue on here no?

Reply:
Ok, got it. Added a silent close version of the API. I think this exception `e` is the main one we want to preserve.